sound/soc/sof/sof-audio.c

```diff
@@ -168,6 +168,306 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler && swidget->pipeline_id == pipeline_id)
+			return swidget;
+
+	return NULL;
+}
+
+int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int pipe_widget_count = 0;
+	int i, ret;
+	int j;
+
+	/* nothing to set up */
+	if (!list)
+		return 0;
+
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget, *temp_pipe_widget;
+		bool pipeline_found = false;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			goto widget_free;
+
+		/* find pipeline widget for the pipeline that this widget belongs to */
+		pipe_widget = sof_find_pipeline_by_id(sdev, swidget->pipeline_id);
+		if (!pipe_widget) {
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		/* Add it to the spcm pipeline list if it isn't part of it already. */
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
+		if (!pipeline_found) {
+			/*
+			 * Pipeline widgets can only be set up after the scheduling
+			 * widget has been set up. Create a temp widget to save it to the spcm
+			 * pipeline list as the original widget is already part of the core's
+			 * widget list.
+			 */
+			temp_pipe_widget = kmemdup(pipe_widget, sizeof(*pipe_widget), GFP_KERNEL);
+			if (!temp_pipe_widget) {
+				ret = -ENOMEM;
+				goto widget_free;
+			}
+
+			list_add(&temp_pipe_widget->list, &spcm->stream[dir].pipeline_list);
+		}
+	}
+
+	/*
+	 * error in setting pipeline connections will result in route status being reset for
+	 * routes that were successfully set up when the widgets are freed.
+	 */
+	ret = sof_setup_pipeline_connections(sdev, list, dir);
+	if (ret < 0)
+		goto widget_free;
+
+	/* setup pipeline widgets and complete pipelines */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = sof_find_widget_by_comp_id(sdev, w->comp_id);
+
+		ret = sof_widget_setup(sdev, pipe_widget);
+		if (ret < 0)
+			goto pipe_free;
+		else
+			pipe_widget_count++;
+
+		if (pipe_widget->complete)
+			continue;
+
+		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
+		if (pipe_widget->complete < 0) {
+			ret = pipe_widget->complete;
+			goto pipe_free;
+		}
+	}
+
+	return 0;
+
+pipe_free:
+	j = 0;
+
+	/* free the pipe widgets that were setup successfully set up */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = sof_find_widget_by_comp_id(sdev, w->comp_id);
+
+		if (j == pipe_widget_count)
```

Comment:
Would it not be better to have a state variable in the sof_widget to track this so we dont have to rely on count. This state could also be dumped in debugFS to help developers (alongside the other SOF info).